pkg/bmc/virtualmedia_validation.go

```diff
@@ -0,0 +1,94 @@
+package bmc
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/stmcginnis/gofish"
+	"github.com/stmcginnis/gofish/redfish"
+)
+
+// contains tells whether the array of redfish.VirtualMediaType
+// has the element redfish.VirtualMediaType.
+func contains(array []redfish.VirtualMediaType, element redfish.VirtualMediaType) bool {
+	for _, n := range array {
+		if element == n {
+			return true
+		}
+	}
+	return false
+}
+
+// checkVirtualMedia tells if all virtual media of a given Manager
+// has the correct type and actions to work with virtual media boot.
+func checkVirtualMedia(m *redfish.Manager) error {
+	vmediaList, err := m.VirtualMedia()
+	if err != nil {
+		return err
+
+	}
+	if len(vmediaList) == 0 {
+		return fmt.Errorf("no virtual media found for manager %s", m.ID)
+	}
+
+	var vmediaError error
+	for _, vmedia := range vmediaList {
+		if !contains(vmedia.MediaTypes, redfish.CDMediaType) {
+			vmediaError = fmt.Errorf("virtual media %s from manager %s does not contain "+
+				"CD MediaType", vmedia.ID, m.ID)
+			continue
+		}
+		if !vmedia.SupportsMediaEject || !vmedia.SupportsMediaInsert {
+			vmediaError = fmt.Errorf("virtual media %s from manager %s does not support "+
+				"at least one of following Actions: InsertMedia, EjectMedia", vmedia.ID, m.ID)
+			continue
+		}
+		return nil
+	}
+	return vmediaError
+}
+
+// ValidateVirtualMedia returns nil when the BMC has support
+// for VirtualMedia and error when the BMC does not have support.
+func ValidateVirtualMedia(address string, user string, password string, insecure bool, systemURI string, httpClient *http.Client) error {
+	bmcConfig := gofish.ClientConfig{
+		Endpoint:   address,
+		Username:   user,
+		Password:   password,
+		Insecure:   insecure,
+		HTTPClient: httpClient,
+	}
+
+	client, err := gofish.Connect(bmcConfig)
+	if err != nil {
+		if strings.Contains(err.Error(), "i/o timeout") {
+			return &AccessVirtualMediaValidationError{message: "timeout reached when trying to access redfish BMC"}
+		}
+		return &AccessVirtualMediaValidationError{message: "could not connect to redfish BMC"}
+	}
+	defer client.Logout()
+
+	system, err := redfish.GetComputerSystem(client, systemURI)
+	if err != nil {
+		return fmt.Errorf("%s not found", systemURI)
+	}
+
+	if len(system.ManagedBy) == 0 {
+		return fmt.Errorf("system %s is not managed by any manager", system.ID)
+	}
+
+	var validationError error
+	for _, availableManager := range system.ManagedBy {
+		manager, err := redfish.GetManager(client, availableManager)
+		if err != nil {
+			validationError = fmt.Errorf("could not access manager %s for system %s",
+				manager.ID, system.ID)
+			continue
+		}
+		validationError = checkVirtualMedia(manager)
```

Comment:
We only need one suitable manager, so once `err == nil` we should return `nil`.

Reply:
We can only return `nil` if the manager we found has the support for virtual media

Reply:
let me know if this makes sense to I can mark as resolved.

Reply:
You're checking a list of managers. What if (purely hypothetically, but still) only the 1st one is suitable? Ironic will work in this case, your code won't.

Reply:
Oh I see, going to push a fix for this!